File: Part_1/Arrays/dynamic_arrays/leetcode/concatenation_of_array.cpp
#include <iostream>
#include <vector>

std::vector<int> getConcatenation(std::vector<int> &nums)
{
  std::vector<int> ans(nums.size() * 2);

  for (size_t i = 0; i < nums.size(); i++)
  {
    ans[i] = nums[i]; 
    ans[i + nums.size()] = nums[i];
  }

  return ans;
}

std::vector<int>& getConcatenationRefrence(std::vector<int> &nums)
{
  std::vector<int>* ans = new std::vector<int>(nums.size() * 2);

  for (size_t i = 0; i < nums.size(); i++)
  {
    (*ans)[i] = nums[i]; 
    (*ans)[i + nums.size()] = nums[i];
  }

  return *ans;
}

int main()
{
  std::vector<int> nums = {1, 2, 1};

  // Receive a reference to the heap-allocated vector
  std::vector<int>& result = getConcatenationRefrence(nums);

  // Print the concatenated vector
  for (int val : result)
  {
    std::cout << val << " ";
  }

  delete &result;

  return 0;
}

File: Part_1/Arrays/dynamic_arrays/leetcode/concatenation_of_array.py
from typing import List

def getConcatenation_V2(nums: List[int]) -> List[int]:
  ans = []
  
  for _ in range(2):
    for n in nums:
      ans.append(n)
    
  return ans


def getConcatenation(nums: List[int]) -> List[int]:
  ans = [0] * (2 * len(nums))
  
  for n in range(len(nums)):
    ans[n] = nums[n]
    ans[len(nums) + n] = nums[n]
    
  return ans


nums = [1, 2, 1]
result = getConcatenation_V2(nums)

print(result)